sdk/core/core-amqp/src/requestResponseLink.ts

```diff
@@ -273,3 +228,75 @@ export class RequestResponseLink implements ReqResLink {
     return new RequestResponseLink(session, sender, receiver);
   }
 }
+
+type NormalizedInfo = {
+  statusCode: number;
+  statusDescription: string;
+  errorCondition: string;
+};
+
+// Handle different variations of property names in responses emitted by EventHubs and ServiceBus.
+const getCodeDescriptionAndError = (props: any): NormalizedInfo => {
+  if (!props) props = {};
+  return {
+    statusCode: (props[Constants.statusCode] || props.statusCode) as number,
+    statusDescription: (props[Constants.statusDescription] || props.statusDescription) as string,
+    errorCondition: (props[Constants.errorCondition] || props.errorCondition) as string
+  };
+};
+
+/**
+ * This is used as the onMessage handler for the "message" event on the receiver.
+ *
+ * (This is inspired from the message settlement sequence in service-bus SDK which
+ * relies on a single listener for settled event for all the messages.)
+ * The sequence is as follows:
+ * 1. User calls `await RequestResponseLink.sendRequest()`
+ * 2. This creates a `Promise` that gets stored in the _responsesMap
+ * 3. When the service acknowledges the response, this method gets called for that request.
+ * 4. We resolve() the promise from the _responsesMap with the message.
+ * 5. User's code after the sendRequest continues.
+ *
+ * @internal
+ * @ignore
+ */
+export function onMessageReceived(
+  context: EventContext,
+  connection: Connection,
+  responsesMap: Map<string, DeferredPromiseWithCallback>
+): void {
+  const message = context.message;
+  if (message) {
+    const responseCorrelationId = message.correlation_id;
+    if (responsesMap.has(responseCorrelationId as string)) {
+      const promise = responsesMap.get(
+        responseCorrelationId as string
+      ) as DeferredPromiseWithCallback;
+      promise.cleanupBeforeResolveOrReject();
+
+      const info = getCodeDescriptionAndError(message.application_properties);
+      if (info.statusCode > 199 && info.statusCode < 300) {
+        logger.verbose(`Resolving the response with correlation-id: ${responseCorrelationId}`);
+        const deleteResult = responsesMap.delete(responseCorrelationId as string);
```

Comment:
Need to `responsesMap.delete` here but not in the 'else' below?

Reply:
Great catch! It should happen for both if and else.
Updated.